### PR TITLE
Dygray/typescript streaming

### PR DESCRIPTION
**Pull request by DylanDmitri**
Support for typescript. It uses a custom `UploadStream` object rather than C# channels. This mostly means that there's no queue on the client side, and that putting a message into the "pipe" before invoking results in an error. Makes the serialization cleaner though.

Also sneaking in a small change that allows streaming invocations to take channel parameters. There's an example that I used for manual verification, and I'm working on unit tests and will push a commit soon™.

### CODE REVIEW

File: clients/ts/signalr/src/HubConnection.ts [hunk 1]
```diff
@@ -509,4 +525,37 @@ export class HubConnection {
             type: MessageType.CancelInvocation,
         };
     }
+
+    public createStreamItem(id: string, item: any): ParameterStreamMessage {
```
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply clients/ts/signalr/src/HubConnection.ts]: rename pl0x

File: clients/ts/signalr/src/IHubProtocol.ts [hunk 2]
```diff
@@ -90,6 +94,18 @@ export interface StreamItemMessage extends HubInvocationMessage {
...
 
+/** A hub message representing a single stream item, transferred through a streaming parameter. */
+export interface ParameterStreamMessage extends HubMessageBase {
+    /** @inheirtDoc */
+    readonly type: MessageType.ParameterStream;
+
+    /** The streamId */
+    readonly streamId: string;
+
+    /** The item produced by the server */
```
Comment (BrennanConroy) [inline clients/ts/signalr/src/IHubProtocol.ts]: not server

File: clients/ts/signalr/tests/HubConnection.test.ts [hunk 3]
```diff
@@ -288,6 +288,39 @@ describe("HubConnection", () => {
             });
         });
 
+        it("is able to send stream items to server", async () => {
```
Comment (BrennanConroy) [inline clients/ts/signalr/tests/HubConnection.test.ts]: single e2e test at least!

File: samples/ClientSample/UploadSample.cs [hunk 4]
```diff
@@ -18,72 +19,147 @@ internal static void Register(CommandLineApplication app)
             {
                 cmd.Description = "Tests a streaming invocation from client to hub";
 
-                var baseUrlArgument = cmd.Argument("<BASEURL>", "The URL to the Chat Hub to test");
+                CommandArgument baseUrlArgument = cmd.Argument("<BASEURL>", "The URL to the Chat Hub to test");
```
Comment (BrennanConroy) [inline samples/ClientSample/UploadSample.cs]: vars

File: samples/SignalRSamples/wwwroot/index.html [hunk 5]
```diff
@@ -22,5 +22,10 @@ <h1>ASP.NET Core SignalR (Streaming)</h1>
         <li><a href="streaming.html?transport=ServerSentEvents">Server Sent Events</a></li>
         <li><a href="streaming.html?transport=WebSockets">Web Sockets</a></li>
     </ul>
+    <h1> ASP.NET Core SignalR (Parameter Streaming via Channels) </h1>
+    <u1>
+        <li><a href="channelParameters.html?transport=ServerSentEvents">SSE</a></li>
```
Comment (BrennanConroy) [inline samples/SignalRSamples/wwwroot/index.html]: Why you h8 longpolling m8

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs [hunk 6]
```diff
@@ -1013,6 +1016,11 @@ private void OnServerTimeout()
 
         private async Task PingServer()
         {
+            if (Debugger.IsAttached)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs]: Remove?

File: test/Microsoft.AspNetCore.SignalR.Tests/HubConnectionHandlerTests.cs [hunk 7]
```diff
@@ -2600,6 +2599,46 @@ public async Task UploadStreamCompleteWithError()
             }
         }
 
+        [Fact]
+        public async Task UploadStreamReturnsDownloadStream()
```
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.Tests/HubConnectionHandlerTests.cs]: "CanPassStreamingParameterToStreamHubMethod"

File: clients/ts/signalr/src/HubConnection.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (DylanDmitri) [reply clients/ts/signalr/src/HubConnection.ts]: can't make private it's used in UploadStream

File: clients/ts/signalr/src/HubConnection.ts [hunk 8]
```diff
@@ -509,4 +525,37 @@ export class HubConnection {
...
+
+    public async complete(error?: string): Promise<void> {
+        if (error) {
+            return this.connection.sendWithProtocol({ type: MessageType.StreamComplete, streamId: this.streamId, error });
+        } else {
+            return this.connection.sendWithProtocol({ type: MessageType.StreamComplete, streamId: this.streamId });
+        }
+    }
+
+    public asPlaceholder() {
```
[the remark this answers is not in the document]
Comment (DylanDmitri) [reply clients/ts/signalr/src/HubConnection.ts]: D E L E T E D  E N T I R E L Y

File: clients/ts/signalr/src/HubConnection.ts [hunk 9]
```diff
@@ -219,6 +231,12 @@ export class HubConnection {
     public invoke<T = any>(methodName: string, ...args: any[]): Promise<T> {
         const invocationDescriptor = this.createInvocation(methodName, args, false);
 
+        // for (let i = 0; i < args.length; i++) {
```
[the remark this answers is not in the document]
Comment (DylanDmitri) [reply clients/ts/signalr/src/HubConnection.ts]: D E L E T E D

File: clients/ts/signalr/src/HubConnection.ts [hunk 10]
```diff
@@ -509,4 +525,37 @@ export class HubConnection {
...
+    public async complete(error?: string): Promise<void> {
+        if (error) {
+            return this.connection.sendWithProtocol({ type: MessageType.StreamComplete, streamId: this.streamId, error });
+        } else {
+            return this.connection.sendWithProtocol({ type: MessageType.StreamComplete, streamId: this.streamId });
+        }
+    }
+
+    public asPlaceholder() {
+        return {streamId: this.streamId};
```
[the remark this answers is not in the document]
Comment (DylanDmitri) [reply clients/ts/signalr/src/HubConnection.ts]: O   k       a     y

File: clients/ts/signalr/src/HubConnection.ts [hunk 11]
```diff
@@ -1,11 +1,9 @@
-import { UploadStream } from "./UploadStream";
-
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 import { HandshakeProtocol, HandshakeRequestMessage, HandshakeResponseMessage } from "./HandshakeProtocol";
 import { IConnection } from "./IConnection";
-import { CancelInvocationMessage, CompletionMessage, IHubProtocol, InvocationMessage, MessageType, StreamDataMessage, StreamInvocationMessage, StreamItemMessage } from "./IHubProtocol";
+import { CancelInvocationMessage, CompletionMessage, IHubProtocol, InvocationMessage, MessageType, StreamCompleteMessage, StreamDataMessage, StreamInvocationMessage, StreamItemMessage } from "./IHubProtocol";
```
Comment (mikaelm12) [inline clients/ts/signalr/src/HubConnection.ts]: We need to add this new message to the HubProtocol spec

File: clients/ts/signalr/src/HubConnection.ts [hunk 12]
```diff
@@ -538,4 +594,27 @@ export class HubConnection {
             type: MessageType.CancelInvocation,
         };
     }
+
+    private createStreamDataMessage(id: string, item: any): StreamDataMessage {
```
Comment (mikaelm12) [inline clients/ts/signalr/src/HubConnection.ts]: Remind me why we need this over a stream item?

File: clients/ts/signalr/src/HubConnection.ts [hunk 13]
```diff
@@ -538,4 +594,27 @@ export class HubConnection {
...
+        return {
+            item,
+            streamId: id,
+            type: MessageType.StreamData,
+        };
+    }
+
+    private createStreamCompleteMessage(id: string, error?: string): StreamCompleteMessage {
```
Comment (mikaelm12) [inline clients/ts/signalr/src/HubConnection.ts]: Or this over just a Completion?

File: clients/ts/signalr/src/HubConnection.ts [hunk 14]
```diff
@@ -508,25 +522,67 @@ export class HubConnection {
...
+                // Replace the stream with a placeholder
+                args[i] = { streamId };
+            }
+        }
+
+        return streams;
+    }
+
+    private isObservable(arg: any): arg is IStreamResult<any> {
+        return arg.subscribe && typeof arg.subscribe === "function";
```
Comment (mikaelm12) [inline clients/ts/signalr/src/HubConnection.ts]: nit: This seems fragile.

**Review by mikaelm12 — COMMENTED** [review]
Seems like some overlap between new message types and existing ones. I think I just lost all context on the design decisions from the summer

File: clients/ts/signalr/src/HubConnection.ts (hunk 12 above)
Comment (BrennanConroy) [reply clients/ts/signalr/src/HubConnection.ts]: Stream Id

File: clients/ts/signalr/src/HubConnection.ts (hunk 14 above)
Comment (BrennanConroy) [reply clients/ts/signalr/src/HubConnection.ts]: Welcome to TS/JS

Comment (anurse) [reply clients/ts/signalr/src/HubConnection.ts]: Yeah, this is totally the legit JS was to do this. It's not pretty, but it works :)

**Comment by BrennanConroy** [comment]
Closing as we moved the PR to https://github.com/aspnet/AspNetCore/pull/4300

**Comment by davidfowl** [comment]
😢

**Pull request closed without merging**